File: leetcode/Generate_Parentheses.cpp
/**
 * Solution: 
 *      l: #(    r: #)
 *  If #( is less than half, then you can keep adding (
 *  If l > r, then you can keep adding ) 
 */
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        if(n == 1) return {"()"};
        
        vector<string> result;
        string sofar = "(";
        
        parenthesis_helper(result, sofar, 1, 0, 2*n);
        return result;
    }
private:
    void parenthesis_helper(vector<string> &result, string sofar, int l, int r, int m){
        if(l+r == m){
            result.push_back(sofar);
            return;
        }
        
        if(l < m/2) parenthesis_helper(result, sofar + "(", l+1, r, m);
        if(l > r) parenthesis_helper(result, sofar + ")", l, r+1, m);
    }
};

File: leetcode/java/GenerateParentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        if(n < 0) throw new IllegalArgumentException("xx");
        if(n == 0) return Collections.emptyList();
        
        List<String> result = new ArrayList<>();
        gph(result, "", 0, 0, n);
        return result;
    }
    
    private void gph(List<String> result, String sofar, int l , int r, int n) {
        if(sofar.length() == 2*n) {
            result.add(sofar);
            return;
        }
        if(l < n) gph(result, sofar+"(", l+1, r, n);
        if(l > r) gph(result, sofar+")", l, r+1, n);
    }
}


class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        gph(result, "", n, n);
        return result;
    }
    
    private void gph(List<String> result, String sofar, int l, int r) {
        if (r == 0) {
            result.add(sofar);
            return;
        }
        if(l > 0) gph(result, sofar+"(", l-1, r);
        if(l < r) gph(result, sofar+")", l, r-1);
    }
}
